### DISCUSSION TITLE
[Feature Request] [storage] Download files with customer-supplied/managed encryption keys

### DISCUSSION THREAD

**Discussion by ksilz** [firebase/firebase-js-sdk#4438] [General]
I [asked about this on Stack Overflow](https://stackoverflow.com/q/65993276/3574) but got no reply in four days. I created a [companion request for FlutterFire](https://github.com/FirebaseExtended/flutterfire/discussions/4954).

My app stores files with Firebase Storage. I want to encrypt these files conveniently on the server. Firebase Storage uses Google Cloud Storage. And Google Cloud Storage offers two options for this: [Customer-supplied encryption keys](https://cloud.google.com/storage/docs/encryption/customer-supplied-keys), where the app provides a key, and [customer-managed encryption keys](https://cloud.google.com/storage/docs/encryption/customer-managed-keys), where the app provides the name of the server-side “encryption service account” in Google cloud storage.

I think my Java back-end that creates my files would be fine: The [Firebase Admin SDK uses the Java Cloud Storage library](https://firebase.google.com/docs/storage/admin/start#google_cloud_storage_client_libraries). And there [Storage.BlobTargetOption](https://googleapis.dev/java/google-cloud-storage/latest/com/google/cloud/storage/Storage.BlobTargetOption.html#method.summary) has an `encryptionKey()` method for the customer-supplied encryption key, and a `kmsKeyName()` method for the customer-managed encryption keys.

But I don’t see how I can download files with customer-supplied/managed encryption keys in the Firebase Javascript SDK. I can't specify a key or key name when [creating the reference to the file](https://github.com/firebase/firebase-js-sdk/blob/40a702eb2c92f37fd7bf3a18502d7d34cf4927a1/packages/storage/src/service.ts#L101). And I can't specify these [when getting the file's download URL](https://github.com/firebase/firebase-js-sdk/blob/40a702eb2c92f37fd7bf3a18502d7d34cf4927a1/packages/storage/src/reference.ts#L361), either.

So I suggest as a feature that Firebase Storage in the Firebase Javascript SDK supports both the customer-supplied & customer-managed encryption keys for Google Cloud Storage. As for the implementation of that feature, getting a download URL could be the place to specify either a customer-supplied encryption key or the name of a customer-managed encryption key.

**Comment by hsubox76** [comment]
In order to track this feature request, can you create a feature request issue? If you pick "bug report" from the options, the form should have some instructions (start the title of the issue with "FR:") and it will be tagged as a feature request. We don't have any formal means to track topics in Github Discussions so this could get lost over time.

**Reply by ksilz** [reply]
Will do! The AngularFire team [told me to create a discussion](https://github.com/angular/angularfire/issues/2752#issuecomment-774727764), though.

**Reply by ksilz** [reply]
@hsubox76 [Done](https://github.com/firebase/firebase-js-sdk/issues/4443).

**Reply by hsubox76** [reply]
Oh, heh, I think David might have had the discussions url saved in his clipboard because we are in the midst of the modularized Firebase alpha, and this is the center for non-bug feedback on that. Thanks for making the FR, I've tagged it as `storage` and we can track it there.

**Reply by ksilz** [reply]
Thank you!

**Reply by samtstern** [reply]
Is this mostly an SDK feature or a backend feature?  Backend feature requests for FIrebase are best reported here:
https://firebase.google.com/support/troubleshooter/report

**Reply by ksilz** [reply]
It's an SDK feature request.

The backend already has it: [Customer-supplied encryption keys](https://cloud.google.com/storage/docs/encryption/customer-supplied-keys), where the app provides a key, and [customer-managed encryption keys](https://cloud.google.com/storage/docs/encryption/customer-managed-keys), where the app provides the name of the server-side “encryption service account” in Google cloud storage

The Java SDK has it, too: [Storage.BlobTargetOption](https://googleapis.dev/java/google-cloud-storage/latest/com/google/cloud/storage/Storage.BlobTargetOption.html#method.summary) has an `encryptionKey()` method for the customer-supplied encryption key, and a `kmsKeyName()` method for the customer-managed encryption keys.
